File: atp-tour/src/main/java/com/silab/atptour/service/PlayerService.java
package com.silab.atptour.service;

import com.silab.atptour.entity.Player;
import com.silab.atptour.exceptions.AtpEntityNotFoundException;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

/**
 * Represents a service containing all the logic for player data management
 * 
 * @author Lazar
 */
public interface PlayerService {

    /**
     * Adds a new player to the database
     * 
     * @param player A {@link Player} object to be added
     * 
     * @return An added {@link Player}
     */
    public Player addPlayer(Player player);

    /**
     * Updates player data in the database
     * 
     * @param player A {@link Player} object to be updated
     * 
     * @return An updated {@link Player}
     * 
     * @throws AtpEntityNotFoundException If the player doesn't exist in the database
     */
    public Player updatePlayer(Player player) throws AtpEntityNotFoundException;

    /**
     * Gets the player with the given id from the database
     * 
     * @param id A long representing player's id
     * 
     * @return A {@link Player} with the given id
     * 
     * @throws AtpEntityNotFoundException If there's no {@link Player} with the given id
     */
    public Player getPlayer(long id) throws AtpEntityNotFoundException;

    /**
     * Gets players from the database. Supports pagination
     * 
     * @param firstName A string representing player's first name
     * @param lastName A string representing player's last name
     * @param birthCountry A string representing name of the player's birth country
     * @param pageable An instance of {@link  Pageable} interface used for pagination
     * 
     * @return A {@link Page} of players
     */
    public Page<Player> getAllPlayers(String firstName, String lastName, String birthCountry, Pageable pageable);
}

File: atp-tour-angular/src/app/core/services/player.service.ts
import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';
import { Player } from 'src/app/models/player.model';
import { environment } from 'src/environments/environment';

/**
 * Service for player data management
 * 
 * @author Lazar
 */
@Injectable({
  providedIn: 'root'
})
export class PlayerService {

  private API_URL: string = environment.API_URL;

  /**
   * @constructor
   * 
   * @param {HttpClient} httpClient 
   */
  constructor(private httpClient: HttpClient) { }

  /**
   * Adds new player to the database
   * 
   * @param {Player} player Player to be added
   * 
   * @returns {Observable<Player>} Added player
   */
  public addPlayer(player: Player): Observable<Player> {
    return this.httpClient.post<Player>(this.API_URL + "/player", player);
  }

  /**
   * Updates player in the database
   * 
   * @param {Player} player Player to be updated
   * 
   * @returns {Observable<Player>} Updated player
   */
  public updatePlayer(player: Player): Observable<Player> {
    return this.httpClient.put<Player>(this.API_URL + "/player", player);
  }

  /**
   * Gets all players from the database
   * 
   * @returns {Observable<Player[]>} Found players
   */
  public getPlayers(): Observable<Player[]> {
    return this.httpClient.get<Player[]>(this.API_URL + "/player")
  }

 /**
  * Filters players based on their last name
  * 
  * @param {string} value Filter value
  * @param {Player[]} players Players for filtering
  * 
  * @returns {Player[]} Filtered players
  */
  public filterPlayers(value: string | Player, players: Player[]): Player[] {
    const filterValue = (value instanceof Player) ? value.lastName : value;
    return players.filter(option => {
      return option.lastName.toLowerCase().includes(filterValue)
    })
  }
}
